src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/TimeSpanTests.cs

```diff
@@ -0,0 +1,119 @@
+        [InlineData(@"""PT-1M""")]
+        [InlineData(@"""PT-1H""")]
+        [InlineData(@"""P-1D""")]
+        [InlineData(@"""P-1M""")]
+        [InlineData(@"""P-1Y""")]
+        [InlineData(@"""T1S""")]
+        [InlineData(@"""T1M""")]
+        [InlineData(@"""T1H""")]
+        [InlineData(@"""1D""")]
+        [InlineData(@"""1M""")]
+        [InlineData(@"""1Y""")]
+        [InlineData(@"""P1Y2M3D4H5M6""")]
+        [InlineData(@"""P1Y2M3DT4H5M6""")]
+        [InlineData(@"""P1Y2M3DT4H5""")]
+        [InlineData(@"""P1Y2M3DT4""")]
+        [InlineData(@"""P1Y2M3""")]
+        [InlineData(@"""P1Y2""")]
+        [InlineData(@"""P1""")]
+        public static void ReadInvalid(string json) =>
+            Assert.Throws<FormatException>(() => JsonSerializer.Deserialize<TimeSpan>(json));
+
+        [Theory]
+        [MemberData(nameof(ReadCases))]
+        public static void Read(string json, TimeSpan expected) =>
+            Assert.Equal(expected, JsonSerializer.Deserialize<TimeSpan>(json));
+
+        [Theory]
+        [MemberData(nameof(WriteCases))]
+        public static void Write(string expected, TimeSpan value) =>
+            Assert.Equal(expected, JsonSerializer.Serialize(value));
+
+        public static IEnumerable<object[]> ReadCases() =>
+            TestCases().Concat(
+                new []
+                {
+                    new object[] { @"""PT1.5S""", Time(0, 0, 1, 5_000_000) },
+                    new object[] { @"""PT1.5M""", Time(0, 1, 30) },
+                    new object[] { @"""PT1.5H""", Time(1, 30, 0) },
+                    new object[] { @"""P1.5D""", Date(0, 0, 1) + Time(12, 0, 0) },
+                    new object[] { @"""P1.5M""", Date(0, 1, 15) },
+                    new object[] { @"""P1.5Y""", Date(1, 6, 0) },
+                    new object[] { @"""PT1,5S""", Time(0, 0, 1, 5_000_000) },
+                    new object[] { @"""PT1,5M""", Time(0, 1, 30) },
+                    new object[] { @"""PT1,5H""", Time(1, 30, 0) },
+                    new object[] { @"""P1,5D""", Date(0, 0, 1) + Time(12, 0, 0) },
+                    new object[] { @"""P1,5M""", Date(0, 1, 15) },
+                    new object[] { @"""P1,5Y""", Date(1, 6, 0) },
+                });
+
+        public static IEnumerable<object[]> WriteCases() =>
+            TestCases().Concat(
+                new []
+                {
+                    new object[] { @"""PT1.5S""", Time(0, 0, 1, 5_000_000) },
+                    new object[] { @"""PT1M30S""", Time(0, 1, 30) },
+                    new object[] { @"""PT1H30M""", Time(1, 30, 0) },
+                    new object[] { @"""P1DT12H""", Date(0, 0, 1) + Time(12, 0, 0) },
+                    new object[] { @"""P1M15D""", Date(0, 1, 15) },
+                    new object[] { @"""P1Y6M""", Date(1, 6, 0) },
+                });
+            
+        private static IEnumerable<object[]> TestCases()
+        {
+            yield return new object[] { @"""PT0S""", TimeSpan.Zero };
```

Comment:
Out of curiosity, what type of format is this? I don't recognize it.

Reply:
It's `duration` from ISO8601. The string represents a time interval of 0 seconds.

Reply:
Interesting, thanks. A few remarks:

* Json.NET serializes TimeSpan values using the standard `D.h:m:s.f` format. Adopting the ISO format means that timespan STJ would not be able to roundtrip newtonsoft values, by default.
* The popular duration.js library has OOTB support for the default .NET format.
* The [XmlConvert](https://docs.microsoft.com/en-us/dotnet/api/system.xml.xmlconvert?view=net-5.0) class already contains an implementation for ISO8601 durations. We should attempt to reuse as much of that as possible. In any case the conversion implementation should be factored out of the json converter itself and tested independently.

Reply:
`XmlConvert` and its internal implementation doesn't support ISO8601 fully and slow. Plus the proposed implementation works with UTF-8 out of the box and does stack allocations only.

@danmoseley, looks like the problem with `TimeSpan` support is that there's no clear understanding how it should work. It would be nice if it can pass an API review process to determine details.

Reply:
> looks like the problem with TimeSpan support is that there's no clear understanding how it should work.

I think adopting ISO durations as the default is ok. It is fairly straightforward to work around it. Ideally though I'd like to see that support being added to the `TimeSpan` class itself rather than being embedded in the converter: the "o" format string is not implemented in `TimeSpan`.

Reply:
Then the logic can be moved to `TimeSpan`, I see no problems with it, if there's a green light.